### PR TITLE
Adds `log4j-docgen` module

### PR DESCRIPTION
**Pull request by ppkarwasz**
Adds a `log4j-docgen` module whose purpose is to automatically generate the documentation of Log4j components based on annotations in the source code and Javadoc comments.

This initial version:

- defines a `/META-INF/log4j/plugins.xml` plugin descriptor, which is similar in nature to Maven's [`/META-INF/maven/plugin.xml`](https://maven.apache.org/ref/3.9.6/maven-plugin-api/plugin.html) and uses the same model generator (Modello),
- provides a `SchemaGenerator` that produces a single XML schema from a collection of `plugins.xml` files.

apache/logging-log4j2#1955

### CODE REVIEW

File: spotbugs-exclude.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<FindBugsFilter
+        xmlns="https://github.com/spotbugs/filter/3.0.0"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0 https://raw.githubusercontent.com/spotbugs/spotbugs/3.1.0/spotbugs/etc/findbugsfilter.xsd">
+  <Match>
+    <Not>
+      <Source name="~target/generated-sources/.*"/>
```
Comment (vy) [inline spotbugs-exclude.xml:24]: Why is this necessary particularly for this project?

File: pom.xml [hunk 2]
```diff
@@ -86,6 +86,7 @@
     <!-- Modules here must have a corresponding entry in `dependencyManagement > dependencies` block below! -->
     <module>log4j-changelog</module>
     <module>log4j-changelog-maven-plugin</module>
+    <module>log4j-docgen</module>
```
Comment (vy) [inline pom.xml:89]: See the comment 3 lines above. Aren't we missing a corresponding entry in `dependencyManagement > dependencies` down below?

File: log4j-tools-parent/.mvn [hunk 3]
```diff
@@ -0,0 +1 @@
+../.mvn
```
Comment (vy) [inline log4j-tools-parent/.mvn:1]: Why do we need these symlinks?

File: log4j-docgen/pom.xml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.xmlunit</groupId>
+      <artifactId>xmlunit-assertj3</artifactId>
+      <version>2.9.1</version>
```
Comment (vy) [inline log4j-docgen/pom.xml:42]:
```suggestion
```

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+import org.apache.logging.log4j.docgen.model.Description;
+import org.apache.logging.log4j.docgen.model.EnumType;
+import org.apache.logging.log4j.docgen.model.EnumValue;
+import org.apache.logging.log4j.docgen.model.PluginAttribute;
+import org.apache.logging.log4j.docgen.model.PluginBundle;
+import org.apache.logging.log4j.docgen.model.PluginElement;
+import org.apache.logging.log4j.docgen.model.PluginEntry;
+import org.apache.logging.log4j.docgen.model.io.stax.PluginBundleStaxReader;
+
+public class DefaultSchemaGenerator implements SchemaGenerator {
```
Comment (vy) [inline log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java:43]: Do we expect `SchemaGenerator` to be extensible? How? In the first release? If not, I would prefer keeping it simple and moving the entire logic to `SchemaGenerator` instead.

**Comment by vy** [comment]
@ppkarwasz, I find it very difficult to understand what this PR does exactly and, maybe more importantly, what it is supposed to deliver. Could you provide some review kit, please? In particular, questions I have:
- How is this tool supposed to be used/run?
- What are its inputs and outputs?
- How does it all work? Modello, `SchemaGenerator`, `configuration.xml`, `plugins.xml`, etc.

I guess, since you are deep in the problem, these might sound obvious to you. But I can assure you that the rest of us have no clue.

File: spotbugs-exclude.xml (hunk 1 above)
Comment (ppkarwasz) [reply spotbugs-exclude.xml:24]: The generated sources contain an XML parser/writer and Spotbugs complains about some `XMLStreamWriter` configuration.

File: log4j-tools-parent/.mvn (hunk 3 above)
Comment (ppkarwasz) [reply log4j-tools-parent/.mvn:1]: I was running Maven from IDEA (Maven view).
The IDE is unable to find `.mvn/jvm.config` in the root project if I launch a sub-module.

Nevertheless I should probably remove these.

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java (hunk 5 above)
Comment (ppkarwasz) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java:43]: The idea behind this was to follow Maven's practice.
They usually create an interface for each service and inject it in the plugin.

File: pom.xml (hunk 2 above)
Comment (ppkarwasz) [reply pom.xml:89]: Fixed

**Comment by ppkarwasz** [comment]
@vy,

Here are some suggestions on how to review this PR.

### How to review this PR

This PR contains multiple contents that might require review:

1. The contents of a `/META-INF/log4j/plugins.xml` file are transformed into Java classes using Modello. These classes are described here: [src/main/mdo/plugins.xml](https://github.com/apache/logging-log4j-tools/blob/f671c034f3d20aef65571456c537b2757e0a9f29/log4j-docgen/src/main/mdo/plugins.xml)

   The model should be minimal and the attributes should be rationally named (naming is not my strong suite). An example of a `/META-INF/log4j/plugins.xml` file is given here: [test/resources/META-INF/log4j/plugins.xml](https://github.com/apache/logging-log4j-tools/blob/f671c034f3d20aef65571456c537b2757e0a9f29/log4j-docgen/src/test/resources/META-INF/log4j/plugins.xml)
   
 2. From the model I generate an XSD. The `SchemaGenerator` accepts multiple `plugins.xml` files to be merged together. There is actually always an _internal_ file that is added: [src/main/resources/org/apache/logging/log4j/docgen/internal/configuration.xml](https://github.com/apache/logging-log4j-tools/blob/f671c034f3d20aef65571456c537b2757e0a9f29/log4j-docgen/src/main/resources/org/apache/logging/log4j/docgen/internal/configuration.xml)
 
    The generated XSD (cf. [src/test/resources/xsd/log4j.xsd](https://github.com/apache/logging-log4j-tools/blob/f671c034f3d20aef65571456c537b2757e0a9f29/log4j-docgen/src/test/resources/xsd/log4j.xsd)) should be powerful enough to easily write a configuration file. Due to the usage of `<group>` each plugin element contains a `<sequence>` of sub-elements, i.e. order matter. If you have a solution that accepts elements in any order let me know (`all` can not contain a `group`).

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java (hunk 5 above)
Comment (vy) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java:43]: How do they inject it? If we are gonna statically link the implementation class at the injection site anyway, what is the advantage of this approach?

Comment (ppkarwasz) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java:43]: The recent way is to use [Sisu](https://eclipse.dev/sisu/), which requires me to add some `@Named` and `@Singleton` annotations.

**Review by vy — APPROVED** [review]
I have some minor comments, but the overall looks good to me. Great job! :100:

## Sanity check

To check if I get the wiring right, I share my notes below.

1. A to-be-implemented `PluginScanner` generates `plugins.xml` (similar to the one in `test/resources/META-INF/log4j/plugins.xml`) from `log4j-core.jar`, `log4j-layout-template-json.jar`, etc.
1. Modello, provided with `src/main/mdo/plugins.xml`, generates DTOs and an XML reader/writer.
1. We use Modello-generated DTOs and the XML reader to read `plugins.xml`.
1. Using DTOs populated from the XML, we manually generate XSD ourselves.

Correct?

### `plugins.xml` file name confusion

There are so many `plugins.xml` and it requires quite some stretch to wrap my mind around them. I would appreciate it if we can use more self-explanatory file names:

- `src/main/mdo/plugins.xml` ⇒ `src/main/mdo/plugins-model.xml`
- `src/test/resources/META-INF/log4j/plugins.xml` ⇒ `sample-plugins.xml`
- `src/test/resources/xsd/log4j.xsd` ⇒ `sample-plugins.xsd`
- `src/main/resources/.../configuration.xml` ⇒ `src/.../supplement-plugins.xml`

### `src/site`

`src/site` is hand-crafted as a showcase, right?

### XML `group`s

I think the problem you're describing is as follows:

```
  <complexType name="org.apache.logging.log4j.core.config.LoggerConfig">
    <sequence>
      <group ref="log4j:org.apache.logging.log4j.core.config.AppenderRef.group" minOccurs="0" maxOccurs="unbounded"/>
      <group ref="log4j:org.apache.logging.log4j.core.Filter.group" minOccurs="0"/>
    </sequence>
    <attribute name="name" type="string"/>
  </complexType>
```

Naively thinking, wouldn't omitting the `sequence` work? That is,

```
  <complexType name="org.apache.logging.log4j.core.config.LoggerConfig">
    <group ref="log4j:org.apache.logging.log4j.core.config.AppenderRef.group" minOccurs="0" maxOccurs="unbounded"/>
    <group ref="log4j:org.apache.logging.log4j.core.Filter.group" minOccurs="0"/>
    <attribute name="name" type="string"/>
  </complexType>
```

File: log4j-docgen/src/main/mdo/plugins.xml [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    <description>Documents a bundle of Log4j plugins.</description>
+    <versionDefinition>
+        <type>field</type>
+        <value>schemaVersion</value>
+    </versionDefinition>
+    <defaults>
+        <default>
+            <key>package</key>
+            <value>org.apache.logging.log4j.docgen.model</value>
+        </default>
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-docgen/src/main/mdo/plugins.xml:32]: Nevermind. Figured that it is where Modello dumps its classes.

File: log4j-docgen/src/main/mdo/plugins.xml [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    </defaults>
+
+    <classes>
+        <class rootElement="true">
+            <name>PluginBundle</name>
+            <description>Documents a bundle of Log4j plugins.</description>
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-docgen/src/main/mdo/plugins.xml:38]: I would probably make more sense to call it `PluginSet` like [Plexus Components](https://codehaus-plexus.github.io/guides/developer-guide/configuration/component-descriptor.html) does.

File: log4j-docgen/src/main/mdo/plugins.xml [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+                </field>
+                <field>
+                    <name>elements</name>
+                    <required>true</required>
+                    <association>
+                        <type>PluginElement</type>
+                        <multiplicity>*</multiplicity>
+                    </association>
+                    <description>List of **all** possible nested components.</description>
+                </field>
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-docgen/src/main/mdo/plugins.xml:160]: Good point,

From a JSON and YAML perspective the distinction is not as evident ("elements" are properties with an object value), but I would keep the distinction for the user: "attributes" are injected by **name**, whereas "elements" are injected by **type**. Knowing that simplifies configuration considerably.

I will look at this again to see if I can simplify things.

File: log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+    private static final String LOG4J_NAMESPACE = "http://logging.apache.org/log4j/2.0/config";
+    private static final String XSD_NAMESPACE = XMLConstants.W3C_XML_SCHEMA_NS_URI;
+    private static final String MULTIPLICITY_UNBOUNDED = "*";
+
+    @Override
+    public void generateSchema(Collection<PluginBundle> bundles, XMLStreamWriter writer) throws XMLStreamException {
+        try {
+            final PluginBundle configurationBundle =
+                    new PluginBundleStaxReader().read(getClass().getResourceAsStream("configuration.xml"));
+            final Set<PluginBundle> extendedBundles = new HashSet<>(bundles);
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/internal/DefaultSchemaGenerator.java:56]: The essential containers should already be `TreeSet`s and `TreeMap`s.

In this case the order shouldn't matter, but I'll check the rest and comment the usages of `HashSet` and `HashMap`.

**Comment by ppkarwasz** [comment]
Hi @vy,

> To check if I get the wiring right, I share my notes below.
> 
>     1. A to-be-implemented `PluginScanner` generates `plugins.xml` (similar to the one in `test/resources/META-INF/log4j/plugins.xml`) from `log4j-core.jar`, `log4j-layout-template-json.jar`, etc.
> 
>     2. Modello, provided with `src/main/mdo/plugins.xml`, generates DTOs and an XML reader/writer.
> 
>     3. We use Modello-generated DTOs and the XML reader to read `plugins.xml`.
> 
>     4. Using DTOs populated from the XML, we manually generate XSD ourselves.
> 
> 
> Correct?

Exactly

> ### `plugins.xml` file name confusion

I corrected it, thanks.

> ### `src/site`
> 
> `src/site` is hand-crafted as a showcase, right?

Yes, I will use it later to test the generation of AsciiDoc. The idea is to know the rough structure of the documentation, so that we know what elements we need.

In the showcase documentation attributes and elements are in separate sections for example, so I would like to distinguish the two types.

> ### XML `group`s
> 
> I think the problem you're describing is as follows:
> 
> ```
>   <complexType name="org.apache.logging.log4j.core.config.LoggerConfig">
>     <sequence>
>       <group ref="log4j:org.apache.logging.log4j.core.config.AppenderRef.group" minOccurs="0" maxOccurs="unbounded"/>
>       <group ref="log4j:org.apache.logging.log4j.core.Filter.group" minOccurs="0"/>
>     </sequence>
>     <attribute name="name" type="string"/>
>   </complexType>
> ```
> 
> Naively thinking, wouldn't omitting the `sequence` work? That is,
> 
> ```
>   <complexType name="org.apache.logging.log4j.core.config.LoggerConfig">
>     <group ref="log4j:org.apache.logging.log4j.core.config.AppenderRef.group" minOccurs="0" maxOccurs="unbounded"/>
>     <group ref="log4j:org.apache.logging.log4j.core.Filter.group" minOccurs="0"/>
>     <attribute name="name" type="string"/>
>   </complexType>
> ```

Unfortunately the XMLSchema XSD does not allow that. I think we'll need to decide whether writing elements in an arbitrary order is important to users. We can assume for now it isn't and improve the schema later.

**Comment by ppkarwasz** [comment]
I am closing this, since the documentation generator needs some refactoring to account for the generation of human documentation. This will be resubmitted in three more digestible parts:

- [ ] A PR for the common infrastructure of schema generation and documentation,
- [ ] the schema generator,
- [ ] the FreeMarker-based documentation generator.
